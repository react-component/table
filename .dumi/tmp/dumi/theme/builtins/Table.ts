// @ts-nocheck
// This file is generated by Umi automatically
// DO NOT CHANGE IT MANUALLY!
export { default } from '/Users/jilin/projects/antd/rc-table/node_modules/_dumi@2.0.17@dumi/theme-default/builtins/Table/index.js';
