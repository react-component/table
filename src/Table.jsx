import React from 'react';
import TableRow from './TableRow';

const Table = React.createClass({
  propTypes: {
    data: React.PropTypes.array,
    expandIconAsCell: React.PropTypes.bool,
    expandOnRowClick: React.PropTypes.bool,
    expandedRowKeys: React.PropTypes.array,
    defaultExpandedRowKeys: React.PropTypes.array,
    useFixedHeader: React.PropTypes.bool,
    columns: React.PropTypes.array,
    prefixCls: React.PropTypes.string,
    bodyStyle: React.PropTypes.object,
    style: React.PropTypes.object,
    rowKey: React.PropTypes.func,
    rowClassName: React.PropTypes.func,
    expandedRowClassName: React.PropTypes.func,
    childrenColumnName: React.PropTypes.string,
    onExpandedRowsChange: React.PropTypes.func,
    indentSize: React.PropTypes.number,
  },

  getDefaultProps() {
    return {
      data: [],
      useFixedHeader: false,
      expandIconAsCell: false,
      expandOnRowClick: false,
      columns: [],
      defaultExpandedRowKeys: [],
      rowKey(o) {
        return o.key;
      },
      rowClassName() {
        return '';
      },
      expandedRowClassName() {
        return '';
      },
      onExpandedRowsChange() {
      },
      prefixCls: 'rc-table',
      bodyStyle: {},
      style: {},
      childrenColumnName: 'children',
      indentSize: 15,
    };
  },

  getInitialState() {
    const props = this.props;
    return {
      expandedRowKeys: props.expandedRowKeys || props.defaultExpandedRowKeys,
      data: this.props.data,
    };
  },

  componentWillReceiveProps(nextProps) {
    if ('data' in nextProps) {
      this.setState({
        data: nextProps.data,
      });
    }
    if ('expandedRowKeys' in nextProps) {
      this.setState({
        expandedRowKeys: nextProps.expandedRowKeys,
      });
    }
  },

  onExpandedRowsChange(expandedRowKeys) {
    if (!this.props.expandedRowKeys) {
      this.setState({
        expandedRowKeys: expandedRowKeys,
      });
    }
    this.props.onExpandedRowsChange(expandedRowKeys);
  },

  onExpanded(expanded, record) {
    const info = this.findExpandedRow(record);
    if (info && !expanded) {
      this.onRowDestroy(record);
    } else if (!info && expanded) {
      const expandedRows = this.getExpandedRows().concat();
      expandedRows.push(this.props.rowKey(record));
      this.onExpandedRowsChange(expandedRows);
    }
  },

  onRowDestroy(record) {
    const expandedRows = this.getExpandedRows().concat();
    const rowKey = this.props.rowKey(record);
    let index = -1;
    expandedRows.forEach((r, i) => {
      if (r === rowKey) {
        index = i;
      }
    });
    if (index !== -1) {
      expandedRows.splice(index, 1);
    }
    this.onExpandedRowsChange(expandedRows);
  },

  getExpandedRows() {
    return this.props.expandedRowKeys || this.state.expandedRowKeys;
  },

  getThs() {
    let ths = [];
    if (this.props.expandIconAsCell) {
      ths.push({
        key: 'rc-table-expandIconAsCell',
        className: `${this.props.prefixCls}-expand-icon-th`,
        title: '',
      });
    }
    ths = ths.concat(this.props.columns);
    return ths.map((c)=> {
      if (c.colSpan !== 0) {
        return <th key={c.key} colSpan={c.colSpan} className={c.className || ''}>{c.title}</th>;
      }
    });
  },

  getExpandedRow(key2, content, visible, className) {
    let key = key2;
    const prefixCls = this.props.prefixCls;
    if (key) {
      key += '-extra-row';
    }
    return (<tr key={key} style={{display: visible ? '' : 'none'}} className={`${prefixCls}-expanded-row ${className}`}>
      {this.props.expandIconAsCell ? <td key="rc-table-expand-icon-placeholder"></td> : ''}
      <td colSpan={this.props.columns.length}>
        {content}
      </td>
    </tr>);
  },

  getRowsByData(data, visible, indent) {
    const props = this.props;
    const columns = props.columns;
    const childrenColumnName = props.childrenColumnName;
    const expandedRowRender = props.expandedRowRender;
    const expandIconAsCell = props.expandIconAsCell;
    const expandOnRowClick = props.expandOnRowClick;
    let rst = [];
    const keyFn = props.rowKey;
    const rowClassName = props.rowClassName;
    const expandedRowClassName = props.expandedRowClassName;
    const needIndentSpaced = props.data.some(record =>
      record[childrenColumnName] && record[childrenColumnName].length > 0);
    for (let i = 0; i < data.length; i++) {
      const record = data[i];
      const key = keyFn ? keyFn(record, i) : undefined;
      const childrenColumn = record[childrenColumnName];
      const isRowExpanded = this.isRowExpanded(record);
      let expandedRowContent;
      if (expandedRowRender && isRowExpanded) {
        expandedRowContent = expandedRowRender(record, i);
      }
      const className = rowClassName(record, i);
      rst.push(<TableRow
        indent={indent}
        indentSize={props.indentSize}
        needIndentSpaced={needIndentSpaced}
        className={className}
        record={record}
        expandIconAsCell={expandIconAsCell}
        expandOnRowClick={expandOnRowClick}
        onDestroy={this.onRowDestroy}
        index={i}
        visible={visible}
        onExpand={this.onExpanded}
        expandable={childrenColumn || expandedRowRender}
        expanded={isRowExpanded}
        prefixCls={`${props.prefixCls}-row`}
        childrenColumnName={childrenColumnName}
        columns={columns}
        key={key}/>);

      const subVisible = visible && isRowExpanded;

      if (expandedRowContent && isRowExpanded) {
        rst.push(this.getExpandedRow(key, expandedRowContent, subVisible, expandedRowClassName(record, i)));
      }
      if (childrenColumn) {
        rst = rst.concat(this.getRowsByData(childrenColumn, subVisible, indent + 1));
      }
    }
    return rst;
  },

  getRows() {
    return this.getRowsByData(this.state.data, true, 0);
  },

  getColGroup() {
    let cols = [];
    if (this.props.expandIconAsCell) {
      cols.push(<col className={`${this.props.prefixCls}-expand-icon-col`} key="rc-table-expand-icon-col"></col>);
    }
    cols = cols.concat(this.props.columns.map((c)=> {
      return <col key={c.key} style={{width: c.width}}></col>;
    }));
    return <colgroup>{cols}</colgroup>;
  },

  findExpandedRow(record) {
    const keyFn = this.props.rowKey;
    const currentRowKey = keyFn(record);
    const rows = this.getExpandedRows().filter((i) => {
      return i === currentRowKey;
    });
    return rows[0] || null;
  },

  isRowExpanded(record) {
    return !!this.findExpandedRow(record);
  },

  render() {
    const props = this.props;
    const prefixCls = props.prefixCls;
    const columns = this.getThs();
    const rows = this.getRows();
    let className = props.prefixCls;
    if (props.className) {
      className += ' ' + props.className;
    }
    let headerTable;
    let thead = (<thead className={`${prefixCls}-thead`}>
    <tr>
      {columns}
    </tr>
    </thead>);
    if (props.useFixedHeader) {
      headerTable = (<div className={`${prefixCls}-header`}>
        <table>
          {this.getColGroup()}
          {thead}
        </table>
      </div>);
      thead = null;
    }
    return (
      <div className={className} style={props.style}>
        {headerTable}
        <div className={`${prefixCls}-body`} style={props.bodyStyle}>
          <table>
            {this.getColGroup()}
            {thead}
            <tbody className={`${prefixCls}-tbody`}>
            {rows}
            </tbody>
          </table>
        </div>
      </div>
    );
  },
});

export default Table;
