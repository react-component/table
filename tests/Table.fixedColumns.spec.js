/* eslint-disable no-undef */
import React from 'react';
import { render, mount } from 'enzyme';
import Table from '../src';

describe('Table.fixedColumns', () => {
  // see:
  // https://github.com/airbnb/enzyme/issues/49#issuecomment-270250193
  // https://github.com/tmpvar/jsdom/issues/653
  function mockClientRect(node, rect) {
    node.getBoundingClientRect = () => ({
      ...rect,
    });
  }
  const columns = [
    { title: 'title1', dataIndex: 'a', key: 'a', width: 100, fixed: 'left' },
    { title: 'title2', dataIndex: 'b', key: 'b', width: 100, fixed: 'left' },
    { title: 'title3', dataIndex: 'c', key: 'c' },
    { title: 'title4', dataIndex: 'b', key: 'd' },
    { title: 'title5', dataIndex: 'b', key: 'e' },
    { title: 'title6', dataIndex: 'b', key: 'f' },
    { title: 'title7', dataIndex: 'b', key: 'g' },
    { title: 'title8', dataIndex: 'b', key: 'h' },
    { title: 'title9', dataIndex: 'b', key: 'i' },
    { title: 'title10', dataIndex: 'b', key: 'j' },
    { title: 'title11', dataIndex: 'b', key: 'k' },
    { title: 'title12', dataIndex: 'b', key: 'l', width: 100, fixed: 'right' },
  ];
  const data = [
    { a: '123', b: 'xxxxxxxx', d: 3, key: '1' },
    { a: 'cdd', b: 'edd12221', d: 3, key: '2' },
    { a: '133', c: 'edd12221', d: 2, key: '3' },
    { a: '133', c: 'edd12221', d: 2, key: '4' },
    { a: '133', c: 'edd12221', d: 2, key: '5' },
    { a: '133', c: 'edd12221', d: 2, key: '6' },
    { a: '133', c: 'edd12221', d: 2, key: '7' },
    { a: '133', c: 'edd12221', d: 2, key: '8' },
    { a: '133', c: 'edd12221', d: 2, key: '9' },
  ];

  it('renders correctly', () => {
    const wrapper = render(<Table columns={columns} data={data} scroll={{ x: 1200 }} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('toggles hover class when user hovering', () => {
    const wrapper = mount(<Table columns={columns} data={data} scroll={{ x: 1200 }} />);
    let tables = wrapper.find('table');
    tables
      .at(0)
      .find('tbody tr')
      .at(0)
      .simulate('mouseEnter');
    tables = wrapper.find('table');
    expect(
      tables
        .at(0)
        .find('tbody tr')
        .at(0)
        .is('.rc-table-row-hover'),
    ).toBe(true);
    expect(
      tables
        .at(1)
        .find('tbody tr')
        .at(0)
        .is('.rc-table-row-hover'),
    ).toBe(true);
    expect(
      tables
        .at(2)
        .find('tbody tr')
        .at(0)
        .is('.rc-table-row-hover'),
    ).toBe(true);
    tables
      .at(0)
      .find('tbody tr')
      .at(0)
      .simulate('mouseLeave');
    tables = wrapper.find('table');
    expect(
      tables
        .at(0)
        .find('tbody tr')
        .at(0)
        .is('.rc-table-row-hover'),
    ).toBe(false);
    expect(
      tables
        .at(1)
        .find('tbody tr')
        .at(0)
        .is('.rc-table-row-hover'),
    ).toBe(false);
    expect(
      tables
        .at(2)
        .find('tbody tr')
        .at(0)
        .is('.rc-table-row-hover'),
    ).toBe(false);
  });

  it('calculate fixedColumns row height', () => {
    const wrapper = mount(<Table columns={columns} data={data} scroll={{ x: 1200 }} />);
    const tableNode = wrapper.instance().tableNode;
    const rows = tableNode.querySelectorAll('tr');
    const theads = tableNode.querySelectorAll('thead');
    const fixedLeftRows = tableNode.querySelectorAll('.rc-table-fixed-left tr');
    const fixedRightRows = tableNode.querySelectorAll('.rc-table-fixed-right tr');
    const rowHeight = '30px';

    function simulateTableShow() {
      mockClientRect(tableNode, {
        top: 0,
        left: 0,
        right: 500,
        bottom: 300,
        width: 500,
        height: 300,
      });
      theads.forEach(thead => {
        mockClientRect(thead, {
          top: 0,
          left: 0,
          right: 500,
          bottom: 30,
          width: 500,
          height: 30,
        });
      });

      const height = parseInt(rowHeight, 10);
      let i = 0;
      rows.forEach(row => {
        i = rows.length / 3 === i ? 0 : i + 1;
        mockClientRect(row, {
          top: i * height,
          left: 0,
          right: 500,
          bottom: (i + 1) * height,
          width: 500,
          height,
        });
      });
    }
    function simulateTableHidden() {
      const rect = {
        top: 0,
        left: 0,
        right: 0,
        bottom: 0,
        width: 0,
        height: 0,
      };
      mockClientRect(tableNode, rect);
      mockClientRect(theads, rect);
      rows.forEach(row => mockClientRect(row, rect));
    }

    // <Table /> is show.
    simulateTableShow();
    wrapper.setProps({});
    fixedLeftRows.forEach(tr => {
      expect(tr.style.height).toBe(rowHeight);
    });
    fixedRightRows.forEach(tr => {
      expect(tr.style.height).toBe(rowHeight);
    });

    // <Table /> is hidden.
    simulateTableHidden();
    wrapper.update();
    fixedLeftRows.forEach(tr => {
      expect(tr.style.height).toBe(rowHeight);
    });
    fixedRightRows.forEach(tr => {
      expect(tr.style.height).toBe(rowHeight);
    });
  });

  it('has correct scroll classNames when table resize', () => {
    const wrapper = mount(
      <Table columns={columns} data={data} scroll={{ x: true }} style={{ width: 2000 }} />,
    );
    const tableNode = wrapper.instance().tableNode;
    const tableBodyContainer = tableNode.querySelectorAll('.rc-table-scroll > .rc-table-body')[0];
    const tableBodyNode = tableBodyContainer.children[0];
    expect(tableNode.className).toContain('rc-table-scroll-position-left');
    expect(tableNode.className).toContain('rc-table-scroll-position-right');
    mockClientRect(tableBodyContainer, {
      width: 500,
    });
    mockClientRect(tableBodyNode, {
      width: 800,
    });
    wrapper.setProps({ style: { width: 500 } });
    expect(tableNode.className).toContain('rc-table-scroll-position-left');
    expect(tableNode.className).not.toContain('rc-table-scroll-position-right');
  });

  it('update fixed column high after columns change', () => {
    const wrapper = mount(<Table columns={[]} data={[]} />);
    wrapper.setProps({ columns, data });
    expect(wrapper.instance().resizeEvent).not.toBeUndefined();
  });
});
