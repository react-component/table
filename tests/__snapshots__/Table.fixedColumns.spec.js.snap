// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table.fixedColumns renders correctly 1`] = `
<div
  class="rc-table rc-table-scroll-position-left"
>
  <div
    class="rc-table-content"
  >
    <div
      class="rc-table-scroll"
    >
      <div
        class="rc-table-body"
        style="overflow-x:scroll;-webkit-transform:translate3d (0, 0, 0)"
      >
        <table
          class="rc-table-fixed"
          style="width:1200px"
        >
          <colgroup>
            <col
              style="width:100px;min-width:100px"
            />
            <col
              style="width:100px;min-width:100px"
            />
            <col />
            <col />
            <col />
            <col />
            <col />
            <col />
            <col />
            <col />
            <col />
            <col
              style="width:100px;min-width:100px"
            />
          </colgroup>
          <thead
            class="rc-table-thead"
          >
            <tr>
              <th
                class="rc-table-fixed-columns-in-body"
              >
                title1
              </th>
              <th
                class="rc-table-fixed-columns-in-body"
              >
                title2
              </th>
              <th
                class=""
              >
                title3
              </th>
              <th
                class=""
              >
                title4
              </th>
              <th
                class=""
              >
                title5
              </th>
              <th
                class=""
              >
                title6
              </th>
              <th
                class=""
              >
                title7
              </th>
              <th
                class=""
              >
                title8
              </th>
              <th
                class=""
              >
                title9
              </th>
              <th
                class=""
              >
                title10
              </th>
              <th
                class=""
              >
                title11
              </th>
              <th
                class="rc-table-fixed-columns-in-body"
              >
                title12
              </th>
            </tr>
          </thead>
          <tbody
            class="rc-table-tbody"
          >
            <tr
              class="rc-table-row rc-table-row-level-0"
              data-row-key="1"
            >
              <td
                class="rc-table-fixed-columns-in-body"
              >
                <span
                  class="rc-table-row-indent indent-level-0"
                  style="padding-left:0px"
                />
                123
              </td>
              <td
                class="rc-table-fixed-columns-in-body"
              >
                xxxxxxxx
              </td>
              <td
                class=""
              />
              <td
                class=""
              >
                xxxxxxxx
              </td>
              <td
                class=""
              >
                xxxxxxxx
              </td>
              <td
                class=""
              >
                xxxxxxxx
              </td>
              <td
                class=""
              >
                xxxxxxxx
              </td>
              <td
                class=""
              >
                xxxxxxxx
              </td>
              <td
                class=""
              >
                xxxxxxxx
              </td>
              <td
                class=""
              >
                xxxxxxxx
              </td>
              <td
                class=""
              >
                xxxxxxxx
              </td>
              <td
                class="rc-table-fixed-columns-in-body"
              >
                xxxxxxxx
              </td>
            </tr>
            <tr
              class="rc-table-row rc-table-row-level-0"
              data-row-key="2"
            >
              <td
                class="rc-table-fixed-columns-in-body"
              >
                <span
                  class="rc-table-row-indent indent-level-0"
                  style="padding-left:0px"
                />
                cdd
              </td>
              <td
                class="rc-table-fixed-columns-in-body"
              >
                edd12221
              </td>
              <td
                class=""
              />
              <td
                class=""
              >
                edd12221
              </td>
              <td
                class=""
              >
                edd12221
              </td>
              <td
                class=""
              >
                edd12221
              </td>
              <td
                class=""
              >
                edd12221
              </td>
              <td
                class=""
              >
                edd12221
              </td>
              <td
                class=""
              >
                edd12221
              </td>
              <td
                class=""
              >
                edd12221
              </td>
              <td
                class=""
              >
                edd12221
              </td>
              <td
                class="rc-table-fixed-columns-in-body"
              >
                edd12221
              </td>
            </tr>
            <tr
              class="rc-table-row rc-table-row-level-0"
              data-row-key="3"
            >
              <td
                class="rc-table-fixed-columns-in-body"
              >
                <span
                  class="rc-table-row-indent indent-level-0"
                  style="padding-left:0px"
                />
                133
              </td>
              <td
                class="rc-table-fixed-columns-in-body"
              />
              <td
                class=""
              >
                edd12221
              </td>
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class="rc-table-fixed-columns-in-body"
              />
            </tr>
            <tr
              class="rc-table-row rc-table-row-level-0"
              data-row-key="4"
            >
              <td
                class="rc-table-fixed-columns-in-body"
              >
                <span
                  class="rc-table-row-indent indent-level-0"
                  style="padding-left:0px"
                />
                133
              </td>
              <td
                class="rc-table-fixed-columns-in-body"
              />
              <td
                class=""
              >
                edd12221
              </td>
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class="rc-table-fixed-columns-in-body"
              />
            </tr>
            <tr
              class="rc-table-row rc-table-row-level-0"
              data-row-key="5"
            >
              <td
                class="rc-table-fixed-columns-in-body"
              >
                <span
                  class="rc-table-row-indent indent-level-0"
                  style="padding-left:0px"
                />
                133
              </td>
              <td
                class="rc-table-fixed-columns-in-body"
              />
              <td
                class=""
              >
                edd12221
              </td>
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class="rc-table-fixed-columns-in-body"
              />
            </tr>
            <tr
              class="rc-table-row rc-table-row-level-0"
              data-row-key="6"
            >
              <td
                class="rc-table-fixed-columns-in-body"
              >
                <span
                  class="rc-table-row-indent indent-level-0"
                  style="padding-left:0px"
                />
                133
              </td>
              <td
                class="rc-table-fixed-columns-in-body"
              />
              <td
                class=""
              >
                edd12221
              </td>
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class="rc-table-fixed-columns-in-body"
              />
            </tr>
            <tr
              class="rc-table-row rc-table-row-level-0"
              data-row-key="7"
            >
              <td
                class="rc-table-fixed-columns-in-body"
              >
                <span
                  class="rc-table-row-indent indent-level-0"
                  style="padding-left:0px"
                />
                133
              </td>
              <td
                class="rc-table-fixed-columns-in-body"
              />
              <td
                class=""
              >
                edd12221
              </td>
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class="rc-table-fixed-columns-in-body"
              />
            </tr>
            <tr
              class="rc-table-row rc-table-row-level-0"
              data-row-key="8"
            >
              <td
                class="rc-table-fixed-columns-in-body"
              >
                <span
                  class="rc-table-row-indent indent-level-0"
                  style="padding-left:0px"
                />
                133
              </td>
              <td
                class="rc-table-fixed-columns-in-body"
              />
              <td
                class=""
              >
                edd12221
              </td>
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class="rc-table-fixed-columns-in-body"
              />
            </tr>
            <tr
              class="rc-table-row rc-table-row-level-0"
              data-row-key="9"
            >
              <td
                class="rc-table-fixed-columns-in-body"
              >
                <span
                  class="rc-table-row-indent indent-level-0"
                  style="padding-left:0px"
                />
                133
              </td>
              <td
                class="rc-table-fixed-columns-in-body"
              />
              <td
                class=""
              >
                edd12221
              </td>
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class=""
              />
              <td
                class="rc-table-fixed-columns-in-body"
              />
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="rc-table-fixed-left"
    >
      <div
        class="rc-table-body-outer"
        style="-webkit-transform:translate3d (0, 0, 0)"
      >
        <div
          class="rc-table-body-inner"
        >
          <table
            class="rc-table-fixed"
          >
            <colgroup>
              <col
                style="width:100px;min-width:100px"
              />
              <col
                style="width:100px;min-width:100px"
              />
            </colgroup>
            <thead
              class="rc-table-thead"
            >
              <tr>
                <th
                  class=""
                >
                  title1
                </th>
                <th
                  class=""
                >
                  title2
                </th>
              </tr>
            </thead>
            <tbody
              class="rc-table-tbody"
            >
              <tr
                class="rc-table-row rc-table-row-level-0"
                data-row-key="1"
              >
                <td
                  class=""
                >
                  <span
                    class="rc-table-row-indent indent-level-0"
                    style="padding-left:0px"
                  />
                  123
                </td>
                <td
                  class=""
                >
                  xxxxxxxx
                </td>
              </tr>
              <tr
                class="rc-table-row rc-table-row-level-0"
                data-row-key="2"
              >
                <td
                  class=""
                >
                  <span
                    class="rc-table-row-indent indent-level-0"
                    style="padding-left:0px"
                  />
                  cdd
                </td>
                <td
                  class=""
                >
                  edd12221
                </td>
              </tr>
              <tr
                class="rc-table-row rc-table-row-level-0"
                data-row-key="3"
              >
                <td
                  class=""
                >
                  <span
                    class="rc-table-row-indent indent-level-0"
                    style="padding-left:0px"
                  />
                  133
                </td>
                <td
                  class=""
                />
              </tr>
              <tr
                class="rc-table-row rc-table-row-level-0"
                data-row-key="4"
              >
                <td
                  class=""
                >
                  <span
                    class="rc-table-row-indent indent-level-0"
                    style="padding-left:0px"
                  />
                  133
                </td>
                <td
                  class=""
                />
              </tr>
              <tr
                class="rc-table-row rc-table-row-level-0"
                data-row-key="5"
              >
                <td
                  class=""
                >
                  <span
                    class="rc-table-row-indent indent-level-0"
                    style="padding-left:0px"
                  />
                  133
                </td>
                <td
                  class=""
                />
              </tr>
              <tr
                class="rc-table-row rc-table-row-level-0"
                data-row-key="6"
              >
                <td
                  class=""
                >
                  <span
                    class="rc-table-row-indent indent-level-0"
                    style="padding-left:0px"
                  />
                  133
                </td>
                <td
                  class=""
                />
              </tr>
              <tr
                class="rc-table-row rc-table-row-level-0"
                data-row-key="7"
              >
                <td
                  class=""
                >
                  <span
                    class="rc-table-row-indent indent-level-0"
                    style="padding-left:0px"
                  />
                  133
                </td>
                <td
                  class=""
                />
              </tr>
              <tr
                class="rc-table-row rc-table-row-level-0"
                data-row-key="8"
              >
                <td
                  class=""
                >
                  <span
                    class="rc-table-row-indent indent-level-0"
                    style="padding-left:0px"
                  />
                  133
                </td>
                <td
                  class=""
                />
              </tr>
              <tr
                class="rc-table-row rc-table-row-level-0"
                data-row-key="9"
              >
                <td
                  class=""
                >
                  <span
                    class="rc-table-row-indent indent-level-0"
                    style="padding-left:0px"
                  />
                  133
                </td>
                <td
                  class=""
                />
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
    <div
      class="rc-table-fixed-right"
    >
      <div
        class="rc-table-body-outer"
        style="-webkit-transform:translate3d (0, 0, 0)"
      >
        <div
          class="rc-table-body-inner"
        >
          <table
            class="rc-table-fixed"
          >
            <colgroup>
              <col
                style="width:100px;min-width:100px"
              />
            </colgroup>
            <thead
              class="rc-table-thead"
            >
              <tr>
                <th
                  class=""
                >
                  title12
                </th>
              </tr>
            </thead>
            <tbody
              class="rc-table-tbody"
            >
              <tr
                class="rc-table-row rc-table-row-level-0"
                data-row-key="1"
              >
                <td
                  class=""
                >
                  xxxxxxxx
                </td>
              </tr>
              <tr
                class="rc-table-row rc-table-row-level-0"
                data-row-key="2"
              >
                <td
                  class=""
                >
                  edd12221
                </td>
              </tr>
              <tr
                class="rc-table-row rc-table-row-level-0"
                data-row-key="3"
              >
                <td
                  class=""
                />
              </tr>
              <tr
                class="rc-table-row rc-table-row-level-0"
                data-row-key="4"
              >
                <td
                  class=""
                />
              </tr>
              <tr
                class="rc-table-row rc-table-row-level-0"
                data-row-key="5"
              >
                <td
                  class=""
                />
              </tr>
              <tr
                class="rc-table-row rc-table-row-level-0"
                data-row-key="6"
              >
                <td
                  class=""
                />
              </tr>
              <tr
                class="rc-table-row rc-table-row-level-0"
                data-row-key="7"
              >
                <td
                  class=""
                />
              </tr>
              <tr
                class="rc-table-row rc-table-row-level-0"
                data-row-key="8"
              >
                <td
                  class=""
                />
              </tr>
              <tr
                class="rc-table-row rc-table-row-level-0"
                data-row-key="9"
              >
                <td
                  class=""
                />
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
